File: js/src/vista-sdk/lib/transport/json/time-series-data/Extensions.ts
import { Defaults, VIS } from "../../..";
import { DataChannelId, ShipId, TimeSeries } from "../../domain";
import { TimeSeriesDto } from "./TimeSeriesData";

export class Extensions {
    public static toJsonDto(
        domain: TimeSeries.TimeSeriesDataPackage
    ): TimeSeriesDto.TimeSeriesDataPackage {
        const p = domain.package;
        const h = domain.package.header;

        return {
            Package: {
                Header: h
                    ? {
                          Author: h.author,
                          DateCreated: h.dateCreated,
                          DateModified: h.dateModified,
                          ShipID: h.shipId.toString(),
                          SystemConfiguration: h.systemConfiguration?.map(
                              (s) => ({
                                  ID: s.id,
                                  TimeStamp: s.timeStamp,
                              })
                          ),
                          TimeSpan: h.timeSpan
                              ? { End: h.timeSpan.end, Start: h.timeSpan.start }
                              : undefined,
                          AdditionalProperties: h.customHeaders,
                      }
                    : undefined,
                TimeSeriesData: p.timeSeriesData.map((t) => ({
                    DataConfiguration: t.dataConfiguration
                        ? {
                              TimeStamp: t.dataConfiguration.timeStamp,
                              ID: t.dataConfiguration.id,
                          }
                        : undefined,
                    EventData: t.eventData
                        ? {
                              NumberOfDataSet: t.eventData.numberOfDataSet,
                              DataSet: t.eventData.dataSet?.map((d) => ({
                                  TimeStamp: d.timeStamp,
                                  DataChannelID: d.dataChannelId.toString(),
                                  Value: d.value,
                                  Quality: d.quality,
                              })),
                          }
                        : undefined,
                    TabularData: t.tabularData?.map((d) => ({
                        NumberOfDataSet: d.numberOfDataSet,
                        NumberOfDataChannel: d.numberOfDataChannel,
                        DataChannelID: d.dataChannelId?.map((i) =>
                            i.toString()
                        ),
                        DataSet: d.dataSet?.map((td) => ({
                            Quality: td.quality,
                            TimeStamp: td.timeStamp,
                            Value: td.value,
                        })),
                    })),
                    AdditionalProperties: t.customProperties,
                })),
            },
        };
    }

    public static async toDomainModel(
        dto: TimeSeriesDto.TimeSeriesDataPackage
    ): Promise<TimeSeries.TimeSeriesDataPackage> {
        const p = dto.Package;
        const h = dto.Package.Header;

        const timeSeriesData: TimeSeries.TimeSeriesData[] = [];

        for (let t of p.TimeSeriesData) {
            const eventDataSet: TimeSeries.EventDataSet[] = [];
            const tabularData: TimeSeries.TabularData[] = [];

            if (t.EventData?.DataSet) {
                for (let ed of t.EventData.DataSet) {
                    const dataChannelId = await DataChannelId.parseAsync(
                        ed.DataChannelID
                    );
                    eventDataSet.push({
                        dataChannelId,
                        timeStamp: ed.TimeStamp,
                        value: ed.Value,
                        quality: ed.Quality,
                    });
                }
            }

            if (t.TabularData) {
                for (let td of t.TabularData) {
                    const dataChannelIds: DataChannelId[] = [];
                    for (let i of td.DataChannelID ?? []) {
                        const dataChannelId = await DataChannelId.parseAsync(i);
                        dataChannelIds.push(dataChannelId);
                    }

                    tabularData.push({
                        numberOfDataSet: td.NumberOfDataSet,
                        numberOfDataChannel: td.NumberOfDataChannel,
                        dataChannelId:
                            dataChannelIds.length > 0
                                ? dataChannelIds
                                : undefined,
                        dataSet: td.DataSet?.map((d) => ({
                            timeStamp: d.TimeStamp,
                            value: d.Value,
                            quality: d.Quality,
                        })),
                    });
                }
            }

            timeSeriesData.push({
                dataConfiguration: t.DataConfiguration && {
                    timeStamp: t.DataConfiguration.TimeStamp,
                    id: t.DataConfiguration.ID,
                },
                customProperties: t.AdditionalProperties,
                eventData: {
                    dataSet: eventDataSet.length > 0 ? eventDataSet : undefined,
                    numberOfDataSet: t.EventData?.NumberOfDataSet,
                },
                tabularData: tabularData.length > 0 ? tabularData : undefined,
            });
        }

        return {
            package: {
                header: h
                    ? {
                          shipId: ShipId.parse(h.ShipID),
                          timeSpan: h.TimeSpan
                              ? { end: h.TimeSpan.End, start: h.TimeSpan.Start }
                              : undefined,
                          dateCreated: h.DateCreated,
                          dateModified: h.DateModified,
                          author: h.Author,
                          systemConfiguration: h.SystemConfiguration?.map(
                              (c) => ({ id: c.ID, timeStamp: c.TimeStamp })
                          ),
                          customHeaders: h.AdditionalProperties,
                      }
                    : undefined,
                timeSeriesData: timeSeriesData,
            },
        };
    }
}

File: csharp/src/Vista.SDK.Apache.Avro/TimeSeriesData/Extensions.cs
using Domain = Vista.SDK.Transport.TimeSeries;

namespace Vista.SDK.Transport.Avro.TimeSeriesData;

public static class Extensions
{
    public static TimeSeriesDataPackage ToAvroDto(this Domain.TimeSeriesDataPackage package)
    {
        var p = package.Package;
        var h = package.Package.Header;
        return new TimeSeriesDataPackage()
        {
            Package = new Package()
            {
                Header = h is null
                    ? null
                    : new Header()
                    {
                        Author = h.Author,
                        DateCreated = h.DateCreated?.DateTime,
                        DateModified = h.DateModified?.DateTime,
                        ShipID = h.ShipId.ToString(),
                        SystemConfiguration = h.SystemConfiguration
                            ?.Select(
                                s =>
                                    new SystemConfiguration()
                                    {
                                        ID = s.Id,
                                        TimeStamp = s.TimeStamp.DateTime
                                    }
                            )
                            .ToList(),
                        TimeSpan = h.TimeSpan is null
                            ? null
                            : new TimeSpan()
                            {
                                Start = h.TimeSpan.Start.DateTime,
                                End = h.TimeSpan.End.DateTime,
                            },
                    },
                TimeSeriesData = p.TimeSeriesData
                    .Select(
                        t =>
                            new TimeSeriesData()
                            {
                                DataConfiguration = t.DataConfiguration is null
                                    ? null
                                    : new DataConfiguration()
                                    {
                                        ID = t.DataConfiguration.Id,
                                        TimeStamp = t.DataConfiguration.TimeStamp.DateTime,
                                    },
                                EventData = t.EventData is null
                                    ? null
                                    : new EventData()
                                    {
                                        DataSet = t.EventData.DataSet
                                            ?.Select(
                                                d =>
                                                    new EventDataSet()
                                                    {
                                                        DataChannelID = d.DataChannelId.ToString(),
                                                        Quality = d.Quality,
                                                        TimeStamp = d.TimeStamp.DateTime,
                                                        Value = d.Value,
                                                    }
                                            )
                                            .ToList(),
                                        NumberOfDataSet = t.EventData.NumberOfDataSet,
                                    },
                                TabularData = t.TabularData
                                    ?.Select(
                                        td =>
                                            new TabularData()
                                            {
                                                DataChannelID = td.DataChannelId
                                                    ?.Select(i => i.ToString())
                                                    .ToList(),
                                                DataSet = td.DataSet
                                                    ?.Select(
                                                        ds =>
                                                            new TabularDataSet()
                                                            {
                                                                Quality = ds.Quality?.ToList(),
                                                                TimeStamp = ds.TimeStamp.DateTime,
                                                                Value = ds.Value.ToList(),
                                                            }
                                                    )
                                                    .ToList(),
                                                NumberOfDataSet = td.NumberOfDataSet,
                                                NumberOfDataChannel = td.NumberOfDataChannel,
                                            }
                                    )
                                    .ToList(),
                            }
                    )
                    .ToList(),
            },
        };
    }

    public static Transport.TimeSeries.TimeSeriesDataPackage ToDomainModel(
        this TimeSeriesDataPackage package
    )
    {
        var p = package.Package;
        var h = package.Package.Header;
        return new Domain.TimeSeriesDataPackage(
            new Domain.Package(
                h is null
                    ? null
                    : new Domain.Header(
                        ShipId.Parse(h.ShipID),
                        h.TimeSpan is null
                            ? null
                            : new Domain.TimeSpan(h.TimeSpan.Start, h.TimeSpan.End),
                        h.DateCreated,
                        h.DateModified,
                        h.Author,
                        h.SystemConfiguration
                            ?.Select(c => new Domain.ConfigurationReference(c.ID, c.TimeStamp))
                            .ToList(),
                        // h.AdditionalProperties.CopyProperties()
                        new Dictionary<string, object>()
                    ),
                p.TimeSeriesData
                    .Select(
                        t =>
                            new Domain.TimeSeriesData(
                                t.DataConfiguration is null
                                    ? null
                                    : new Domain.ConfigurationReference(
                                        t.DataConfiguration.ID,
                                        t.DataConfiguration.TimeStamp
                                    ),
                                t.TabularData
                                    ?.Select(
                                        td =>
                                            new Domain.TabularData(
                                                td.NumberOfDataSet,
                                                td.NumberOfDataChannel,
                                                td.DataChannelID
                                                    .Select(i => DataChannelId.Parse(i))
                                                    ?.ToList(),
                                                td.DataSet
                                                    ?.Select(
                                                        tds =>
                                                            new Domain.TabularDataSet(
                                                                tds.TimeStamp,
                                                                tds.Value.ToList(),
                                                                tds.Quality.ToList()
                                                            )
                                                    )
                                                    .ToList()
                                            )
                                    )
                                    .ToList(),
                                t.EventData is null
                                    ? null
                                    : new Domain.EventData(
                                        t.EventData.NumberOfDataSet,
                                        t.EventData.DataSet
                                            ?.Select(
                                                ed =>
                                                    new Domain.EventDataSet(
                                                        ed.TimeStamp,
                                                        DataChannelId.Parse(ed.DataChannelID),
                                                        ed.Value,
                                                        ed.Quality
                                                    )
                                            )
                                            .ToList()
                                    ),
                                // t.AdditionalProperties.CopyProperties()
                                new Dictionary<string, object>()
                            )
                    )
                    .ToList()
            )
        );
    }
}
